### PR TITLE
Improve performance for Uninstall-PSResource by caching

### PR DESCRIPTION
**Pull request by adityapatwardhan**
# PR Summary

Cache the output of Get-Module -ListAvailable for every invocation of the cmdlet.

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/UninstallPSResource.cs [hunk 1]
```diff
@@ -343,50 +353,64 @@ private bool CheckIfDependency(string pkgName, out ErrorRecord errorRecord)
...
-                {
-                    pkgsWithRequiredModules = results.Where(
-                        pkg => ((ReadOnlyCollection<PSModuleInfo>)pkg.Properties["RequiredModules"].Value).Where(
-                            rm => rm.Name.Equals(pkgName, StringComparison.InvariantCultureIgnoreCase)).Any());
-                }
-                catch (Exception e)
+                _dependentModules = _pwsh.AddCommand("Microsoft.PowerShell.Core\\Get-Module").AddParameter("ListAvailable").Invoke<PSModuleInfo>();
+            }
+
+            // Structure of LINQ call:
```
Comment (alerickson) [inline src/code/UninstallPSResource.cs]: ``
```suggestion
```

File: src/code/UninstallPSResource.cs [hunk 2]
```diff
@@ -343,50 +353,64 @@ private bool CheckIfDependency(string pkgName, out ErrorRecord errorRecord)
...
+            // Results is a collection of PSModuleInfo objects that contain a property listing module dependencies, "RequiredModules".
```
Comment (alerickson) [inline src/code/UninstallPSResource.cs]:
```suggestion
            // _dependentModules is a collection of PSModuleInfo objects that contain a property listing module dependencies, "RequiredModules".

```

File: src/code/UninstallPSResource.cs [hunk 3]
```diff
@@ -343,50 +353,64 @@ private bool CheckIfDependency(string pkgName, out ErrorRecord errorRecord)
...
+                    "UninstallPSResourcePackageIsaDependency",
+                    ErrorCategory.InvalidOperation,
+                    null);
 
-                    return true;
-                }
+                return true;
             }
 
+
```
Comment (alerickson) [inline src/code/UninstallPSResource.cs]:
```suggestion

```

File: src/code/UninstallPSResource.cs [hunk 4]
```diff
@@ -162,6 +166,12 @@ protected override void ProcessRecord()
             }
         }
 
+        protected override void EndProcessing()
+        {
+            _dependentModules?.Clear();
```
Comment (alerickson) [inline src/code/UninstallPSResource.cs]: what's the benefit to clearing _dependentModules here?  If there's pipeline input doesn't ProcessRecord() potentially occur multiple times before EndProcessing() is hit?

**Review by alerickson — CHANGES_REQUESTED** [review]

File: src/code/UninstallPSResource.cs (hunk 4 above)
Comment (adityapatwardhan) [inline src/code/UninstallPSResource.cs]:
```suggestion
```

**Review by alerickson — APPROVED** [review]

**Pull request merged**